src/core/ddsi/src/ddsi_sockwaitset.c

```diff
@@ -281,6 +284,252 @@ int ddsi_sock_waitset_next_event (struct ddsi_sock_waitset_ctx * ctx, struct dds
+      return 0;
+    else
+      n++;
+  }
+
+  if (fidx == UINT32_MAX)
+  {
+    const uint32_t newsz = ddsrt_atomic_add32_nv (&ws->sz, WAITSET_DELTA);
+    ws->entries = ddsrt_realloc (ws->entries, newsz * sizeof (*ws->entries));
+    for (idx = sz; idx < newsz; idx++)
+      ws->entries[idx].fd = -1;
+    fidx = sz;
+  }
+  ev.events = EPOLLIN;
+  ev.data.ptr = &ws->entries[fidx];
+  if (epoll_ctl (ws->epfd, EPOLL_CTL_ADD, fd, &ev) == -1)
+    return -1;
+  ws->entries[fidx].conn = conn;
+  ws->entries[fidx].fd = fd;
+  ws->entries[fidx].index = n;
+  return 1;
+}
+
+struct ddsi_sock_waitset * ddsi_sock_waitset_new (void)
+{
+  const uint32_t sz = WAITSET_DELTA;
+  struct ddsi_sock_waitset * ws;
+  uint32_t i;
+  if ((ws = ddsrt_malloc (sizeof (*ws))) == NULL)
+    goto fail_waitset;
+  ddsrt_atomic_st32 (&ws->sz, sz);
+  if ((ws->entries = ddsrt_malloc (sz * sizeof (*ws->entries))) == NULL)
+    goto fail_entries;
+  for (i = 0; i < sz; i++)
+    ws->entries[i].fd = -1;
+  ws->ctx.nevs = 0;
+  ws->ctx.index = 0;
+  ws->ctx.evs_sz = sz;
+  if ((ws->ctx.evs = ddsrt_malloc (ws->ctx.evs_sz * sizeof (*ws->ctx.evs))) == NULL)
+    goto fail_ctx_evs;
+  if ((ws->epfd = epoll_create (1)) == -1)
+    goto fail_kqueue;
+  if (pipe (ws->pipe) == -1)
+    goto fail_pipe;
+  if (add_entry_locked (ws, NULL, ws->pipe[0]) < 0)
+    goto fail_add_trigger;
+  assert (ws->entries[0].fd == ws->pipe[0]);
+  if (fcntl (ws->epfd, F_SETFD, fcntl (ws->epfd, F_GETFD) | FD_CLOEXEC) == -1)
+    goto fail_fcntl;
+  if (fcntl (ws->pipe[0], F_SETFD, fcntl (ws->pipe[0], F_GETFD) | FD_CLOEXEC) == -1)
+    goto fail_fcntl;
+  if (fcntl (ws->pipe[1], F_SETFD, fcntl (ws->pipe[1], F_GETFD) | FD_CLOEXEC) == -1)
+    goto fail_fcntl;
+  ddsrt_mutex_init (&ws->lock);
+  return ws;
+
+fail_fcntl:
+fail_add_trigger:
+  close (ws->pipe[0]);
+  close (ws->pipe[1]);
+fail_pipe:
+  close (ws->epfd);
+fail_kqueue:
+  ddsrt_free (ws->ctx.evs);
+fail_ctx_evs:
+  ddsrt_free (ws->entries);
+fail_entries:
+  ddsrt_free (ws);
+fail_waitset:
+  return NULL;
+}
+
+void ddsi_sock_waitset_free (struct ddsi_sock_waitset * ws)
+{
+  ddsrt_mutex_destroy (&ws->lock);
+  close (ws->pipe[0]);
+  close (ws->pipe[1]);
+  close (ws->epfd);
+  ddsrt_free (ws->entries);
+  ddsrt_free (ws->ctx.evs);
+  ddsrt_free (ws);
+}
+
+void ddsi_sock_waitset_trigger (struct ddsi_sock_waitset * ws)
+{
+  char buf = 0;
+  int n;
+  n = (int)write (ws->pipe[1], &buf, 1);
+  if (n != 1)
+  {
+    DDS_WARNING("ddsi_sock_waitset_trigger: read failed on trigger pipe, errno = %d\n", errno);
+  }
+}
+
+int ddsi_sock_waitset_add (struct ddsi_sock_waitset * ws, struct ddsi_tran_conn * conn)
+{
+  int ret;
+  ddsrt_mutex_lock (&ws->lock);
+  ret = add_entry_locked (ws, conn, ddsi_conn_handle (conn));
+  ddsrt_mutex_unlock (&ws->lock);
+  return ret;
+}
+
+void ddsi_sock_waitset_purge (struct ddsi_sock_waitset * ws, unsigned index)
+{
+  /* Sockets may have been closed by the Purge is called, but any closed sockets
```

Comment:
Some references to 'kqueue' in the comment (also on line 485), and a few words missing in the sentence '.. by the Purge is called' (also in the original in the kqueue implementation)